### PR TITLE
[BTRX-701] Fix Search Criteria

### PR DESCRIPTION
**Pull request by tadeork**
## Addresses
This PR takes two stories https://broadinstitute.atlassian.net/browse/BTRX-701 and https://broadinstitute.atlassian.net/browse/BTRX-642
We needed to fix first the search response times to be sure that we're testing the statuses properly.

## Changes
-  Added index in `fundings` table.
-  Added new statuses
-  Now search in two columns of `issue` for status `status` and `approval_status`
-  Label in status form search

## Testing
Describe for the reviewer how to test the specifics of this PR, both positive and negative cases.

---

- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Get a thumb from Belatrix
- [ ] **Submitter**: Get a thumb from @rushtong
- [ ] **Submitter**: Merge to develop using github's "Squash and Merge" feature
- [ ] **Submitter**: Delete branch after merge

### CODE REVIEW

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 1]
```diff
@@ -728,7 +729,46 @@ class QueryService implements Status {
...
+        final String query = "SELECT i.id id, " +
+                "i.project_key projectKey, " +
+                "i.type type, " +
+                "i.status status, " +
+                "i.summary summary, " +
+                "i.reporter reporter, " +
+                "i.update_date updated, " +
+                "i.expiration_date expirationDate " +
+                "FROM issue i " +
+                "WHERE i.id IN (" + issueIds.join(",") + ")"
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Is there a limit to the number of ids here? I know that oracle has a 1000 item limit, not sure if mysql or cloudsql has the same kind of thing.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 2]
```diff
@@ -728,7 +729,46 @@ class QueryService implements Status {
...
+            issueSearchItem.setStatus(it.get("status").toString())
+            issueSearchItem.setSummary(it.get("summary").toString())
+            issueSearchItem.setReporter(it.get("reporter").toString())
+            if (it.get("expirationDate") != null) {
+                issueSearchItem.setExpirationDate(it.get("expirationDate"))
+            }
+            if (it.get("updated") != null) {
+                issueSearchItem.setUpdateDate(it.get("updated"))
+            }
+            result.push(issueSearchItem)
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Curious about two things. 
1. Why this is an ordered list? The query is not ordered, but you are forcing the dtos into an array list, so why not just a set?
2. Why use `push`? That's a groovy add-on that just does `add` under the hood - you might have better performance with `add`

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 3]
```diff
@@ -728,7 +729,46 @@ class QueryService implements Status {
...
+                "i.reporter reporter, " +
+                "i.update_date updated, " +
+                "i.expiration_date expirationDate " +
+                "FROM issue i " +
+                "WHERE i.id IN (" + issueIds.join(",") + ")"
+
+        List<Issue> result = new ArrayList<Issue>()
+
+        getSqlConnection().rows(query).each{
+            Issue issueSearchItem = new Issue()
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Why create a new issue here? You could let hibernate give you a populated `Issue` object from the query which would avoid the awkward parsing, e.g. https://github.com/broadinstitute/orsp-pub/blob/feature/grails-app/services/org/broadinstitute/orsp/QueryService.groovy#L314

This is also a risk because it introduces a new, untested, pattern for hydrating objects. I do worry about what Hibernate will do with all of these dirty (unsaved) new objects and whether that introduces a separate set of performance/memory concerns.

Comment (tadeork) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: If I do as the example the search performance drops drastically from seconds up to five minutes. On the other hand we're not doing anything with the Issue objects but showing the results and only the fields that are mapped, there's no need of other info. So we're not using unnecessary memory.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 1 above)
Comment (tadeork) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: MySQL has 1GB of limit https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet
I think this will not be a problema.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 3 above)
Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: The concern isn't that we are doing anything with unnecessary memory, the concern is that Grails/Hibernate is doing something. `Issue` is a domain object which means Grails treats it differently than other classes. It could be the case that we're creating memory leaks this way, but I just don't know for sure which is why I bring up the question.

Another alternative is to create some other kind of object that is not a domain object and pass that along as the list of things to return as search results.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 1 above)
Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: This was my question, so I think there is no real limit to the number of elements in the in-clause: https://stackoverflow.com/questions/4275640/mysql-in-condition-limit/4275704

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 3 above)
Comment (tadeork) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: I don't quite get the concern. The issue objects are read only in this scenario. Another thing, I already try to use a DTO and got very complex very fast. But when I added the index I found that I there was no need to use a DTO because it was very fast and got a lot simpler. Also I found that I was going to have to change the way is used `PermissionService.issueIsForbiden`
In this commit is more or less that approach 87cba6a

Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: This is also one of my concerns: 

> The issue objects are read only in this scenario.

This method assumes that these are read-only, but there is nothing to guarantee that. Imagine if some controller called this method, got a bunch of issues back, and tried to do stuff with them. You'd now have modified `Issue` objects that could be re-saved, and only partially (meaning data will be deleted) because you didn't fully populate them from this method. You've chosen a non-standard way to use Grails domain classes which could lead to real problems.

**Comment by tadeork** [comment]
I've made several changes. I think we need another round of testing from @wloforteBroad to be sure that everything is OK.

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy [hunk 4]
```diff
@@ -10,31 +10,61 @@ class PermissionService implements UserInfo {
     UserService userService
 
     // get issue's collaborators as a List<String>
-    List<String> getIssueCollaborators(Issue issue) {
-      issue.extraProperties.findAll ({ it.name == IssueExtraProperty.COLLABORATOR }).collect { property -> property.value }
+    List<String> getIssueCollaborators(Map<String, List<String>> extraProperties) {
+        List<String> results = new ArrayList<String>()
+
+        def collaboratorsValues = extraProperties.findAll ({ it.key == IssueExtraProperty.COLLABORATOR }).values()
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: nit: avoid `def` when possible. This is one of those times 😄

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy [hunk 5]
```diff
@@ -10,31 +10,61 @@ class PermissionService implements UserInfo {
...
+        if (collaboratorsValues.size() != 0) {
+            results = collaboratorsValues.first()
+        }
+
+        collaboratorsValues = extraProperties.findAll ({ it.key == IssueExtraProperty.COLLABORATORS }).values()
+        if (results.size() == 0 && collaboratorsValues.size() != 0) {
+            results = collaboratorsValues.first()
+        }
+
+        results
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: This whole method is a little confusing. 
* Why aren't the collaborators *all* of the collaborators? 
* Why is it only the first one (in each of the two current cases)? 
* Why do we mutate the results? 
* Why iterate twice over the `extraProperties`? 

It would make more sense to me to have something like this:
```
List<String> getIssueCollaborators(Map<String, List<String>> extraProperties) {
    extraProperties.
        findAll({ [IssueExtraProperty.COLLABORATOR, IssueExtraProperty.COLLABORATORS].contains(it.key) }).
        values()
}
```

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy [hunk 6]
```diff
@@ -10,31 +10,61 @@ class PermissionService implements UserInfo {
...
     // get issue's pms as a List<String>
-    List<String> getIssuePMs(Issue issue) {
-      issue.extraProperties.findAll ({ it.name == IssueExtraProperty.PM }).collect { property -> property.value }
+    List<String> getIssuePMs(Map<String, List<String>> extraProperties) {
+        List<String> results = new ArrayList<String>()
+        def collaboratorsValues = extraProperties.findAll ({ it.key == IssueExtraProperty.PM }).values()
+        if (collaboratorsValues.size() != 0) {
+            results = collaboratorsValues.first()
+        }
+        results
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: Here again, why are we only showing a single one? The method name, return type, and comments indicate very different behavior.

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy [hunk 7]
```diff
@@ -10,31 +10,61 @@ class PermissionService implements UserInfo {
...
     // get issue's pis as a List<String>
-    List<String> getIssuePIs(Issue issue) {
-      issue.extraProperties.findAll ({ it.name == IssueExtraProperty.PI }).collect { property -> property.value }
+    List<String> getIssuePIs(Map<String, List<String>> extraProperties) {
+        List<String> results = new ArrayList<String>()
+        def collaboratorsValues = extraProperties.findAll ({ it.key == IssueExtraProperty.PI }).values()
+        if (collaboratorsValues.size() != 0) {
+            results = collaboratorsValues.first()
+        }
+        results
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: More of the same feedback.

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy [hunk 8]
```diff
@@ -10,31 +10,61 @@ class PermissionService implements UserInfo {
...
+    def issueIsForbidden(IssueSearchItemDTO issue, String userName, boolean isAdmin, boolean isViewer) {
+        Map<String, List<String>> extraProperties = issue.extraProperties
+        userHasIssueAccess(issue.reporter, extraProperties, userName, isAdmin, isViewer)
+    }
+
+    def userHasIssueAccess(String reporter, Map<String, List<String>> extraProperties, String userName, boolean isAdmin, boolean isViewer) {
+        boolean userHasAccess = (reporter == userName
+                || getIssueCollaborators(extraProperties).indexOf(userName) >= 0
+                || getIssuePMs(extraProperties).indexOf(userName) >= 0
+                || getIssuePIs(extraProperties).indexOf(userName) >= 0
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: These boolean checks should be looking at *all* of the PMs, PIs, Collaborators, etc. IndexOf is probably the wrong approach here. I think you want to check `contains` on the list.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 9]
```diff
@@ -725,14 +727,73 @@ class QueryService implements Status {
...
+                "WHERE i.id IN (" + issueIds.join(",") + ")"
+
+        Set<IssueSearchItemDTO> resultDTO = new HashSet<IssueSearchItemDTO>()
+        IssueSearchItemDTO issueSearchItemDTO
+        String currentProjectKey = ""
+
+        getSqlConnection().rows(query).each {
+            if (it.get("projectKey") == currentProjectKey) {
+                if (it.get("type") != IssueType.CONSENT_GROUP.name) {
+                    def extraProp = issueSearchItemDTO.extraProperties.get(it.get("name").toString()) ?
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: avoid `def` here.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 10]
```diff
@@ -725,14 +727,73 @@ class QueryService implements Status {
...
+                }
+                if (it.get("updated") != null) {
+                    issueSearchItemDTO.setUpdateDate(it.get("updated"))
+                }
+                if (it.get("type") != IssueType.CONSENT_GROUP.name) {
+                    def extraProp = issueSearchItemDTO.extraProperties.size() != 0 ? issueSearchItemDTO.extraProperties.get(it.get("name").toString()) : []
+                    extraProp.push(it.get("value").toString())
+                    issueSearchItemDTO.extraProperties.put(it.get("name").toString(), extraProp)
+                }
+            }
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Pretty much *all* of this logic should be in a constructor for `IssueSearchItemDTO` that understands how to construct itself from a sql query row.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 11]
```diff
@@ -725,14 +727,73 @@ class QueryService implements Status {
...
+                    extraProp.push(it.get("value").toString())
+                }
+            } else {
+                if (currentProjectKey != "") {
+                    resultDTO.add(issueSearchItemDTO)
+                }
+                currentProjectKey = it.get("projectKey")
+                issueSearchItemDTO = new IssueSearchItemDTO()
+
+                issueSearchItemDTO.setId((Integer)it.get("id"))
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: I'm concerned about the casting here. We should strive mightily to avoid this. If you put this logic into the constructor, you could use a try block around a `Integer.valueOf()` line.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 12]
```diff
@@ -725,14 +727,73 @@ class QueryService implements Status {
...
+                issueSearchItemDTO.setProjectKey(it.get("projectKey").toString())
+                issueSearchItemDTO.setType(it.get("type").toString())
+                issueSearchItemDTO.setStatus(it.get("status").toString())
+                issueSearchItemDTO.setSummary(it.get("summary").toString())
+                issueSearchItemDTO.setReporter(it.get("reporter").toString())
+                if (it.get("expirationDate") != null) {
+                    issueSearchItemDTO.setExpirationDate(it.get("expirationDate"))
+                }
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Why the if condition? Could be a one liner with safe-null checks:
```
issueSearchItemDTO.setExpirationDate(it.get("expirationDate")?.toString())
```

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 13]
```diff
@@ -725,14 +727,73 @@ class QueryService implements Status {
...
+                if (it.get("updated") != null) {
+                    issueSearchItemDTO.setUpdateDate(it.get("updated"))
+                }
+                if (it.get("type") != IssueType.CONSENT_GROUP.name) {
+                    def extraProp = issueSearchItemDTO.extraProperties.size() != 0 ? issueSearchItemDTO.extraProperties.get(it.get("name").toString()) : []
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Avoid def.

Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Confused by this ... extraProp is either a single string value or an empty list?

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 14]
```diff
@@ -725,14 +727,73 @@ class QueryService implements Status {
...
+                    extraProp.push(it.get("value").toString())
+                    issueSearchItemDTO.extraProperties.put(it.get("name").toString(), extraProp)
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: I'm finding this logic confusing ... can you comment on what's happening here?

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy (hunk 5 above)
Comment (tadeork) [reply grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: It's not the first of the values but the result of this is a array with the Collaborators user names within an array. I need to take out from there the actual values. This suggestion is working but I still need use `.first()`

Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: If that's true, then there's something wrong with how you're constructing the values inside `collaboratorsValues`. It should just be just a simple list of names. If instead it is a list of lists, then the underlying method should be super clear about what it's returning.

Comment (tadeork) [reply grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: I not agree, `Issue`s are being instantiated in some way that the `extraProperties` are in the same way as my DTO. But I'll fix boths.

**Comment by vvicario** [comment]
@rushtong Can you please take a look and let us know if we need to do something else? Thanks!

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 15]
```diff
@@ -725,14 +727,65 @@ class QueryService implements Status {
...
+                "i.type type, " +
+                "i.status status, " +
+                "i.summary summary, " +
+                "i.reporter reporter, " +
+                "i.update_date updated, " +
+                "i.expiration_date expirationDate, " +
+                "iep.* " +
+                "FROM issue i LEFT JOIN issue_extra_property iep " +
+                "ON (iep.project_key = i.project_key AND iep.name in ('pm','pi','collaborator')) " +
+                "WHERE i.id IN (" + issueIds.join(",") + ")"
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: This line is subject to sql injection attacks. This should be a named argument. There are lots of other examples in this class that do this.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 16]
```diff
@@ -725,14 +727,65 @@ class QueryService implements Status {
...
+
+        Set<IssueSearchItemDTO> resultDTO = new HashSet<IssueSearchItemDTO>()
+        IssueSearchItemDTO issueSearchItemDTO
+        String currentProjectKey = ""
+
+        getSqlConnection().rows(query).each {
+            if (it.get("projectKey") == currentProjectKey) {
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: All of this logic should be in a constructor for `IssueSearchItemDTO`. That class should accept a `GroovyRowResult` as an argument (that's what `it` is in this line) and perform all of the logic there. Having this in the query service breaks encapsulation and makes it impossible to re-use `IssueSearchItemDTO` in any other context.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 14 above)
Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Any feedback on this question? This looks confusing enough to warrant some kind of comment that describes what it's doing.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 10 above)
Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: @vvicario - why was this marked as resolved? I still think this comment should be addressed.

File: grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy [hunk 17]
```diff
@@ -105,25 +106,20 @@ final class IssueUtils {
         null
     }
 
-    static Map<String, Object> generateArgumentsForRedirect(Issue issue, String id, String tab) {
+    static Map<String, Object> generateArgumentsForRedirect(String type, String id, String tab) {
         Map<String, Object> arguments = new HashMap<>()
-        if (isConsentGroup(issue)) {
+        if (type == IssueType.CONSENT_GROUP.name) {
```
Comment (rushtong) [inline grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy]: I don't understand why this change (in combination with removing `isConsentGroup()`) was necessary. Are there issues with the wrong controller? Also, this change seems very out of scope for the task of improving search performance.

File: grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy [hunk 18]
```diff
@@ -105,25 +106,20 @@ final class IssueUtils {
...
+            tab != null ? arguments.put("params", [projectKey: id, tab: tab]) :
+                    arguments.put("params", [projectKey: id])
         }
         arguments
     }
-
-    static boolean isConsentGroup(Issue issue) {
-        return issue.type == IssueType.CONSENT_GROUP.name || issue.controller == IssueType.CONSENT_GROUP.name
-    }
-
```
Comment (rushtong) [inline grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy]: See comment above.

File: grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy (hunk 17 above)
Comment (tadeork) [reply grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy]: It seems out of scope but I found out that it was simpler to use this method with a String for the IssueType than sending the Issue. And also, and because I couldn't use this method with IssueSearchItemDTO. While I was figuring out how to use this method with Issue Search DTO I found that this approach was simpler, solve my problem and needed less code.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 16 above)
Comment (tadeork) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Agree.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 19]
```diff
@@ -725,15 +727,58 @@ class QueryService implements Status {
...
+                if (currentProjectKey != "") {
+                    resultDTO.add(issueSearchItemDTO)
+                }
+
+                currentProjectKey = it.get("projectKey")
+                issueSearchItemDTO = new IssueSearchItemDTO(it.toSorted())
 
+                if (it.get("type") != IssueType.CONSENT_GROUP.name) {
+                    issueSearchItemDTO.setExtraProperty(it.get("name").toString(), it.get("value").toString())
+                }
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Note that lines 763-765 are exactly the same as 774-776. These, too, should be in the constructor and not here in the query service. In fact, I would suggest that the entire if-else should be there. If you have custom constructor logic spread out in different places, you lose the ability to rely on the class's ability to create itself outside of this query service context. All knowledge about what an `IssueSearchItemDTO` is should be in the constructor.

**Comment by tadeork** [comment]
This last approach works well but a little slower than the first intent. Any way we think we can improve in the future with pagination (we've to create a ticket to track that) as we did with Fundings, Roles Management using Bootstrap Table 2.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 20]
```diff
@@ -725,15 +726,31 @@ class QueryService implements Status {
...
+        final String query = "SELECT * FROM issue i WHERE i.id IN (:issueIds)"
+        SessionFactory sessionFactory = grailsApplication.getMainContext().getBean('sessionFactory')
+        final session = sessionFactory.currentSession
+        final SQLQuery sqlQuery = session.createSQLQuery(query)
+        final results = sqlQuery.with {
+            addEntity(Issue)
+            setParameterList('issueIds', issueIds)
+            list()
+        }
+        results
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: @tadeork - you are correct here, this will be a little slower than processing the raw `GroovyRowResults` directly. You don't *have* to use hibernate here (you are with the `addEntity` call). I don't know the exact syntax off-hand, but it might be something as simple as 
```
        final results = sqlQuery.with {
            setParameterList('issueIds', issueIds)
            list()
        }
```
And that should give you a list of raw row results without having to cast each one as an `Issue` object (which happens under the hood by GORM/hibernate).

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy [hunk 21]
```diff
@@ -10,31 +10,45 @@ class PermissionService implements UserInfo {
...
     }
 
     // verifies if logged user belongs to some user list ....
-    def issueIsForbidden(Issue issue, String userName, boolean isAdmin, boolean isViewer) {
+    Boolean issueIsForbidden(Issue issue, String userName, boolean isAdmin, boolean isViewer) {
+        Map<String, List<String>> extraProperties = issue.extraPropertiesMap
+        userHasIssueAccess(issue.reporter, extraProperties, userName, isAdmin, isViewer)
+    }
+
+    Boolean issueIsForbidden(IssueSearchItemDTO issue, String userName, boolean isAdmin, boolean isViewer) {
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: Are you even using this any more now that your query returns full `Issue` objects?

File: src/main/groovy/org/broadinstitute/orsp/IssueSearchItemDTO.groovy [hunk 22]
```diff
@@ -0,0 +1,48 @@
+package org.broadinstitute.orsp
+
+import groovy.util.logging.Slf4j
+
+@Slf4j
+class IssueSearchItemDTO {
```
Comment (rushtong) [inline src/main/groovy/org/broadinstitute/orsp/IssueSearchItemDTO.groovy]: This looks unused now.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 20 above)
Comment (tadeork) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: This is not working. I need some how to cast the results to Issues.

Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Something like this should also work - it doesn't use hibernate at all (sqlQuery still uses hibernate to generate a raw sql query ... fast, but not as fast as groovy sql)

```
getSqlConnection().rows(query, ["issueIds": issueIds]).collect { it -> ... make something from the it, which is a GroovyRowResult ... }
```

Comment (tadeork) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: I already tried this approach and also didn't work. It doesn't returns any result. I tried several approaches and it was frustrating bcs only a few ones work but slowly.

Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Yeah, I was running a local instance to test and it does look like that doesn't work. No worries on this topic then.
